File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-7.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int idx;
    unordered_map<int, int> pos;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        idx = 0;
        for (int i = 0; i < inorder.size(); i++) {
            pos[inorder[i]] = i;
        }
        TreeNode* root = buildNode(0, (int)inorder.size() - 1, preorder, inorder);
        return root;
    }
    TreeNode* buildNode(int left, int right, vector<int>& preorder, vector<int>& inorder) {
        if (left > right) {
            return nullptr;
        }
        TreeNode* root = new TreeNode(preorder[idx]);
        int inorder_pos = pos[preorder[idx]];
        idx++;
        root->left = buildNode(left, inorder_pos - 1, preorder, inorder);
        root->right = buildNode(inorder_pos + 1, right, preorder, inorder);
        return root;
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxSum = float("-inf")
        def dfs(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            left = max(dfs(node.left), 0)
            right = max(dfs(node.right), 0)
            self.maxSum = max(self.maxSum, node.val + left + right)
            return node.val + max(left, right)
        _ = dfs(root)
        return self.maxSum